_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,268 @@
+my-cluster-kafka-1                            1/1     Running   0          18m
+my-cluster-kafka-2                            1/1     Running   0          18m
+my-cluster-zookeeper-0                        1/1     Running   0          20m
+my-cluster-zookeeper-1                        1/1     Running   0          20m
+my-cluster-zookeeper-2                        1/1     Running   0          20m
+strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
+```
+
+#### Migrating from ZooKeeper to KRaft-based brokers
+
+The initial migration step involves deploying a `KafkaNodePool` custom resource to provision the KRaft controllers. 
+Then, you change the `strimzi.io/kraft` annotation on the `Kafka` custom resource from `disabled` to `migration` to initiate the process.
+
+Here's a snippet of the `KafkaNodePool` custom resource needed to deploy and activate the KRaft controllers:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - controller
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: false
+```
+
+The corresponding `Kafka` custom resource is then updated to set the `strimzi.io/kraft` annotation to `migration`:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+You can do that by annotating the custom resource.
+
+```shell
+kubectl annotate kafka my-cluster strimzi.io/kraft="migration" --overwrite
+```
+
+The annotation change triggers the Strimzi operator to run the following steps:
+
+* Deploy the KRaft controllers with all the necessary connection details to ZooKeeper and the migration flag enabled.
+* Configure the running brokers with the connection details to the KRaft controller quorum, with the migration flag enabled, and initiate a rolling update.
+* During each reconciliation, monitor the migration status by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric. Update the `Kafka` custom resource to reflect this status through the `status.kafkaMetadataState` field.
+* Once the migration is complete, reconfigure the brokers to disconnect from ZooKeeper, disable the migration flag, and initiate another rolling update.
+
+As you can see, just applying one single annotation value has covered most of the phases in the migration procedure with the user doing practically nothing, just following what the operator is running for them.
+
+While the migration is going on, you can see the cluster changing the metadata state during the migration process from `ZooKeeper` to `KRaftPostMigration`, and then waiting for the user to finalize it.
+
+```shell
```

Comment:
Maybe it would be good to have the command here as well? Just to show how the users can watch the state change (if you think it's a good idea, this applies to other places as well)?

Reply:
I will put the command just here, because the other places are snippet from the same output while the migration is running. I mean the end the user runs `kubectl get kafka my-cluster -n kafka -w` just here at the beginning (not every time).